File: 2021/day1.py
import unittest


def part_one(input_text: str | None = None) -> int:
    with open(input_text if input_text else "input/day1_input", encoding="utf-8") as f:
        nums = [int(line.strip()) for line in f]
        return sum(a < b for a, b in zip(nums, nums[1:]))


def part_two(input_text: str | None = None) -> int:
    with open(input_text if input_text else "input/day1_input", encoding="utf-8") as f:
        nums = [int(line.strip()) for line in f]
        return sum(a < b for a, b in zip(nums, nums[3:]))


class Test(unittest.TestCase):
    def test_part_one(self):
        self.assertEqual(part_one("input/day1_test_input"), 7)

    def test_part_two(self):
        self.assertEqual(part_two("input/day1_test_input"), 5)


if __name__ == "__main__":
    print(part_one())
    print(part_two())

File: 2021/src/bin/day1.rs
use std::fs;

fn part_one(text: &str) -> u32 {
    let values: Vec<u32> = text.lines().map(|v| v.parse().unwrap()).collect();
    let mut result = 0;
    for i in 1..values.len() {
        if values[i] > values[i - 1] {
            result += 1;
        }
    }
    result
}

fn part_two(text: &str) -> u32 {
    let values: Vec<u32> = text.lines().map(|v| v.parse().unwrap()).collect();
    let mut result = 0;
    for i in 0..values.len() - 3 {
        if values[i..i + 3].iter().sum::<u32>() < values[i + 1..i + 4].iter().sum::<u32>() {
            result += 1;
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let lines = fs::read_to_string("input/day1_test_input").expect("Can't read file");
        assert_eq!(part_one(&lines), 7);
    }

    #[test]
    fn test_part_two() {
        let lines = fs::read_to_string("input/day1_test_input").expect("Can't read file");
        assert_eq!(part_two(&lines), 5);
    }
}

fn main() {
    let lines = fs::read_to_string("input/day1_input").expect("Can't read file");
    println!("{}", part_one(&lines));
    println!("{}", part_two(&lines));
}
